Fix `audio_get_status` result

kira-bruneau opened the pull request:
`audio_get_status` returns an audio status, and doesn't propagate the transmit result:

- https://github.com/Pulse-Eight/libcec/blob/libcec-6.0.2/src/libcec/LibCECC.cpp#L539-L545
- https://github.com/Pulse-Eight/libcec/blob/libcec-6.0.2/src/libcec/LibCEC.cpp#L572-L575
- https://github.com/Pulse-Eight/libcec/blob/libcec-6.0.2/src/libcec/CECClient.cpp#L824-L829
- https://github.com/Pulse-Eight/libcec/blob/libcec-6.0.2/src/libcec/devices/CECAudioSystem.cpp#L153-L171

Also the status doesn't neatly fit into a Rust enum, so I removed the existing enum and added a wrapper struct.

![image](https://github.com/user-attachments/assets/4c4f0fc2-e0ef-4c8f-8c04-e7de39d8b3ba)

Page 260 of [High-Definition Multimedia Interface Specification 1.3a](https://web.archive.org/web/20171009194844/http://www.microprocessor.org/HDMISpecification13a.pdf)

<!--
Please, make sure:
- you have read the contributing guidelines:
  https://github.com/ssalonen/cec-rs/blob/master/docs/CONTRIBUTING.md
- you have formatted the code using rustfmt:
  https://github.com/rust-lang/rustfmt
- you have checked that all tests pass, by running `cargo test --all`
- you have updated the changelog (if needed):
  https://github.com/ssalonen/cec-rs/blob/master/CHANGELOG.md
-->

kira-bruneau committed:
Fix audio_get_status result

ssalonen commented:
Thanks!

Hmm the enum*.rs is autogenerated by https://github.com/ssalonen/libcec-sys/blob/master/scripts/run-bindgen/bindgen.sh

So somethings needs to be done there... need to have a bit closer look to comment something sensible

can you add simple tests to the conversion similar to other such functions in this lib?

kira-bruneau commented:
Oh ok! I'll just add them back. I mainly just removed them because the new struct made those enums unnecessary.

I'll also keep using the constants from `libcec_sys` for the struct, because `CEC_AUDIO_VOLUME_STATUS_UNKNOWN` conflicts with `CEC_AUDIO_VOLUME_STATUS_MASK`. It'd be a bit confusing to use`CecAudioStatus::MuteStatusMask` to check when the status is unknown.

ssalonen commented:
Yes I think it's ok to keep all enums as-is and having this usability-layer separately like you have done here with new struct

kira-bruneau commented:
I refactored my changes a bit so this new struct can be used with https://github.com/ssalonen/cec-alsa-sync.

- Replaced `KnownCecAudioStatus::new` with a method that takes a separate `volume` & `is_muted` parameter.
- Added a `TryFrom<u8>` implementation to handle converting from a raw status (also adds a variant for the reserved value error case)

ssalonen commented on src/lib.rs:
```diff
@@ -134,6 +135,124 @@ mod util_tests {
+    }
+
+    pub fn is_muted(self) -> bool {
+        self.0 & libcec_sys::CEC_AUDIO_MUTE_STATUS_MASK as u8 != 0
```
Would linting break with parentheses?
```suggestion
        (self.0 & libcec_sys::CEC_AUDIO_MUTE_STATUS_MASK as u8) != 0
```

ssalonen commented on src/lib.rs:
```diff
@@ -134,6 +135,124 @@ mod util_tests {
+    }
+
+    pub fn volume(self) -> u8 {
+        self.0 & libcec_sys::CEC_AUDIO_VOLUME_STATUS_MASK as u8
```
This is probably just me but I am having reading the precedence of operations

for me this verbose version is easier on the eyes. What do you think? does lint accept ir?

```suggestion
        self.0 & (libcec_sys::CEC_AUDIO_VOLUME_STATUS_MASK as u8)
```

kira-bruneau replied:
Oh yep! I can add parenthesis. I'm just used to knowing that casts have higher precedence than all other binary expressions: https://doc.rust-lang.org/reference/expressions.html#expression-precedence

ssalonen commented on src/lib.rs:
```diff
@@ -134,6 +135,124 @@ mod util_tests {
+                Err(Self::Error::Unknown)
+            } else {
+                Err(Self::Error::Reserved(
+                    volume - libcec_sys::CEC_AUDIO_VOLUME_MAX as u8 - 1,
```
```suggestion
                    volume - (libcec_sys::CEC_AUDIO_VOLUME_MAX as u8) - 1,
```

ssalonen reviewed (commented):
Thank you! Added few small colments

ssalonen committed:
Merge branch 'master' into fix-audio-get-status-result

ssalonen committed:
lint fixes for  lib.rs

ssalonen committed:
Update lib.rs

ssalonen committed:
Merge branch 'master' into fix-audio-get-status-result

ssalonen committed:
Update lib.rs
